components/service/glean/src/main/java/mozilla/components/service/glean/Glean.kt

```diff
@@ -129,11 +130,13 @@ open class GleanInternalAPI internal constructor () {
         // If Glean is being initialized from a test, then we want to skip running the
         // startupCheck() as it can cause some intermittent test failures due to multi-
         // threaded race conditions.
-        @Suppress("EXPERIMENTAL_API_USAGE")
         if (!Dispatchers.API.testingMode) {
             metricsPingScheduler.startupCheck()
         }
 
+        // Signal Dispatcher that init is complete
+        Dispatchers.API.flushQueuedInitialTasks()
```

Comment:
So, I understand the underlying data structure is threadsafe and lock free, but what happens if we are iterating over it to clear it and another thread is adding to it?  Would those events fall on the floor?  Should we add a lock to synchronize against that?

Reply:
I guess that wouldn't be possible, since Glean.initialized is set to true above, so maybe it's ok.

Trying to think about how one could write a test for this that wouldn't be flaky...

Reply:
The actual diversion from launch going into the queue is stopped before it processes the queue at all.  That way, all calls to launch _after_ `queueInitialTasks` is set to false would just be processed normally.  

So, by the time we start processing the queue, coroutines should be executing like normal run mode and it shouldn't be possible to add any more to the queue because the `queueInitialTasks` flag is no longer true.

Reply:
That makes sense.  Thanks for bearing with me.  Seems like a risky change wrt multithreading, but I don't have any specific thing to point at...

Reply:
As for setting up a way to test this?  Let me know if you come up with something as I couldn't come up with a reliable way to test the edge case of adding while the queue is processing, aside from putting some very long running tasks in the queue (but again, no way to guarantee that the queue was still processing when adding new tasks)

Reply:
I was also concerned about how this would affect the multi-threading balance we have going for us, but after testing this more and more, I think using the threadsafe collection and the way that it diverts the tasks to just running when it gets ready to process the queue makes me feel pretty confident about it.  The only thing I can see that might happen is that tasks that get added while it's processing the queue could get recorded before tasks that were added to the queue causing something to potentially get recorded out of chronological order.

Reply:
> The only thing I can see that might happen is that tasks that get added while it's processing the queue could get recorded before tasks that were added to the queue causing something to potentially get recorded out of chronological order.

Trying to think of where this might matter.  I'm optimistic it doesn't matter much...

- Timespans (if `stop` and `start`) were unordered -- however, `start` is actually synchronous, so I think that's a non-issue.
- String Lists -- that's just an argument for moving to String Sets which we're already planning
- Events shouldn't matter, since everything is timestamped, we can always sort by the timestamp if it matters for a particular analysis.  (I think the assumption already is that the *could* be out-of-order).
- Explicit calls to `sendPing` -- the ping may not include recent recordings that otherwise would be in the ping.  But I guess they'll show up in the next ping, so ultimately not an issue...?

Reply:
I see your concern, it's a good one. At the very least we should call this out with an extensive comment in `Dspatchers.flushQueuedInitialTasks`.

Synchronizing this would make us pay the cost of synchronization even *after* init. I'm afraid I don't have a good solution to offer for this off the top of my head. Can we also file a bug about this?

Reply:
I filed [this bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1568503) to follow up on this issue.